L80-csharp-load-balancing.md

```diff
@@ -186,37 +272,37 @@ namespace Grpc.Net.Client.Balancer {
         public void Dispose();
         public abstract void RequestConnection();
         public abstract void UpdateChannelState(ChannelState state);
-        public abstract void UpdateSubchannelState(Subchannel subchannel, SubchannelState state);
         protected virtual void Dispose(bool disposing);
     }
     public abstract class LoadBalancerFactory {
         protected LoadBalancerFactory();
         public abstract string Name { get; }
-        public abstract LoadBalancer Create(IConnectionController controller, IDictionary<string, object> options);
+        public abstract LoadBalancer Create(IChannelControlHelper controller, IDictionary<string, object> options);
     }
-    public class PickContext {
+    public sealed class PickContext {
         public PickContext();
         public HttpRequestMessage? Request { get; set; }
     }
-    public class PickFirstBalancer : LoadBalancer {
-        public PickFirstBalancer(IConnectionController channel, ILoggerFactory loggerFactory);
+    public sealed class PickFirstBalancer : LoadBalancer {
+        public PickFirstBalancer(IChannelControlHelper controller, ILoggerFactory loggerFactory);
         public override void RequestConnection();
         public override void UpdateChannelState(ChannelState state);
-        public override void UpdateSubchannelState(Subchannel subchannel, SubchannelState state);
         protected override void Dispose(bool disposing);
     }
-    public class PickFirstBalancerFactory : LoadBalancerFactory {
+    public sealed class PickFirstBalancerFactory : LoadBalancerFactory {
         public PickFirstBalancerFactory(ILoggerFactory loggerFactory);
         public override string Name { get; }
-        public override LoadBalancer Create(IConnectionController controller, IDictionary<string, object> options);
+        public override LoadBalancer Create(IChannelControlHelper controller, IDictionary<string, object> options);
     }
-    public class PickResult {
+    public sealed class PickResult {
         public Status Status { get; }
         public Subchannel? Subchannel { get; }
-        public static PickResult ForError(Status status);
-        public static PickResult ForNoResult();
-        public static PickResult ForSubchannel(Subchannel subchannel, Action<CompleteContext>? onComplete = null);
-        public void Complete(CompleteContext context);
+        public PickResultType Type { get; }
+        public static PickResult ForDrop(Status status);
+        public static PickResult ForFailure(Status status);
+        public static PickResult ForQueue();
+        public static PickResult ForSubchannel(Subchannel subchannel, Action<CompletionContext>? onComplete = null);
```

Comment:
I'll write some tests to verify what happens.

Reply:
> How will onComplete work in the case where the pick is abandoned because the returned subchannel was not in READY state?

onComplete should always be called. The context includes an error property that indicates if there was an HTTP error while making the call.

The .NET HTTP API doesn't provide low-level information about whether a stream was created or not, so a counter will need to be incremented when the pick of a subchannel happens, and is always decremented in onComplete. If the picked subchannel changes from READY to another state before the HTTP request is started then the call will fail and onComplete will run and decrement the count.

Reply:
I noticed the "Least used" policy proposal being reviewed at https://github.com/grpc/proposal/pull/269

.NET is an implementation that would need to increment the counter before stream creation.

Reply:
In the case that I mentioned in https://github.com/grpc/proposal/pull/240/files#r744134047, the channel should abandon the result that was returned by the picker, because it is unusable.  In that case, it is not actually starting a call, so incrementing the counter is not going to yield an accurate result.

For cases like circuit breaking and the least-request policy, what you are describing here is sub-optimal but probably workable; you will increment the counter earlier than you should, but you will immediately decrement it again when `onComplete` runs.  But for cases like load reporting, the LB policy needs some way to know that the pick was abandoned, or else it will report a call that never actually happened.

Reply:
For implementations that increment the counter before stream creation, won't there always be a race? There will always be a small gap in time after incrementing that counter where the picked subchannel goes from a ready state to something else.

Unfortunately, the .NET HTTP API doesn't have an event to indicate a stream was started. Incrementing before stream creation is the only option so we'll have to be sub-optimal. I'll call out this limitation in the documentation.

Reply:
I don't think it's necessary to actually create the HTTP/2 stream on the wire before incrementing the counter.  The race here isn't about creating the stream on the wire; it's actually about getting the transport from the subchannel.

There are three different views of the subchannel's state here:
- The LB policy's view of the subchannel.
- The subchannel's view of its transport.
- The transport's view of the underlying connection.

Each of those will have its own synchronization, so they will get out of sync with respect to each other.  So there are two different races that can happen here:
1. The LB policy thinks the subchannel is in state READY and returns it for a pick, but by the time the channel sees the pick result and asks the subchannel for its transport, the subchannel no longer has a transport.
2. When the channel asks the subchannel for its transport, the subchannel may return the transport, and the channel will start the call on that transport.  But by the time the transport goes to actually put that call on the wire, the underlying connection may have failed.

In case 2, the only thing you can do is fail the call.  We are actually using the pick result in that case, so it's fine to have incremented any counters by the time we get to that point.  When the call fails, we will report that failure to the LB policy, and it can decrement any counters it needs to.  From the LB policy's perspective, this is not really any different from any other call failure.

The case I'm talking about here is case 1.  In that situation, the LB policy has told the channel to start a call on a particular subchannel, but the channel cannot actually start the call, because the subchannel's transport disappeared between when the LB policy returned the subchannel and when the channel tried to get the transport from that subchannel.  In that situation, the call is never actually started, so the pick is basically abandoned: the call gets queued and will wait for a new picker to be returned before being reattempted.  In this case, the ideal behavior is never to have incremented any counters to begin with.

The ideal way to do this is to have an API like Java or C-core, where the channel invokes the tracer's `Start()` method after obtaining the transport from the subchannel, and the LB policy uses `Start()` to increment its counter.  That way, if the pick is abandoned because the transport could not be obtained, `Start()` will never be called, so the counter won't have been incremented to begin with.

The Go API is less optimal, because there is no `Start()` method, so the picker needs to increment the counter before it knows whether or not the pick will be abandoned, and it will need to decrement the counter when `onComplete()` is called.  But at least there, `onComplete()` is passed an indication of whether or not the call was abandoned, so the LB policy has the option of adjusting its data accordingly.  (For example, if the LB policy has a count of calls started and a count of calls finished, it can handle abandoned calls by decrementing calls-started *without* incrementing calls-finished.  This is still sub-optimal, because both counters may have been used to generate a load report between when the picker incremented the counter and when `onComplete()` was called, but at least the data will be eventually consistent.)

I think if you have only an `onComplete()` callback without passing it any indication of whether the call was abandoned, then you have no way to tell the LB policy that the call was never sent, so it can't even correct its data after the fact.  This means that you are always going to report bad data when you abandon a pick.

Reply:
I added `ISubchannelCallTracker`. See the most recent commit for the exact API change.

Reply:
Thanks, the `ISubchannelCallTracker` approach looks much better!